FIAT/hdiv_trace.py

```diff
@@ -49,47 +54,96 @@ class HDivTrace(FiniteElement):
+
+            assert len(ref_el.cells) == len(degree), (
+                "Number of specified degrees must be equal "
+                "to the number of cells."
+            )
+        else:
+            if not ref_el.get_shape() in [TRIANGLE, TETRAHEDRON,
+                                          QUADRILATERAL]:
+                raise NotImplementedError(
+                    "Trace element on a %s not implemented" % type(ref_el)
+                )
+            # Cannot have varying degrees for these reference cells
+            if isinstance(degree, tuple):
+                raise ValueError(
+                    "Must have a tensor product cell "
+                    "if providing multiple degrees"
+                )
+
+        # Initialize entity dofs and construct the DG elements
+        # for the facets
+        facet_sd = sd - 1
+        dg_elements = {}
         entity_dofs = {}
-
-        # Looping over dictionary of cell topology to construct the empty
-        # dictionary for entity ids of the trace element
         topology = ref_el.get_topology()
-        for top_dim, entities in topology.items():
+        for top_dim, entities in iteritems(topology):
+            cell = ref_el.construct_subelement(top_dim)
             entity_dofs[top_dim] = {}
+
+            # We have a facet entity!
+            if cell.get_spatial_dimension() == facet_sd:
+                dg_elements[top_dim] = construct_dg_element(cell, degree)
+            # Initialize
             for entity in entities:
                 entity_dofs[top_dim][entity] = []
 
-        # Filling in entity ids and generating points for dual basis
-        nf = dglagrange.space_dimension()
-        points = []
-        num_facets = sd + 1
-        for f in range(num_facets):
-            entity_dofs[sd - 1][f] = range(f * nf, (f + 1) * nf)
+        # Compute the dof numbering for all facet entities
+        # and extract nodes
+        offset = 0
+        pts = []
+        for facet_dim in sorted(dg_elements):
+            element = dg_elements[facet_dim]
+            nf = element.space_dimension()
+            num_facets = len(topology[facet_dim])
 
-            for dof in dglagrange.dual_basis():
-                facet_point = list(dof.get_point_dict().keys())[0]
-                transform = ref_el.get_entity_transform(sd - 1, f)
-                points.append(tuple(transform(facet_point)))
+            for i in range(num_facets):
+                entity_dofs[facet_dim][i] = list(range(offset + i * nf,
+                                                       offset + (i + 1) * nf))
+
+                # Run over nodes and collect the points for point evaluations
+                for dof in element.dual_basis():
+                    facet_pt, = dof.get_point_dict()
+                    transform = ref_el.get_entity_transform(facet_dim, i)
+                    pts.append(tuple(transform(facet_pt)))
+
+            offset += nf * num_facets
```

Comment:
I think this is now correct, however, I prefer the pattern used [here](https://github.com/firedrakeproject/fiat/pull/8/files?diff=split#diff-9934bc5ebd94879d69bba736b9883070R264).